### PR TITLE
Refactor of Atmos unit tests, Part I

### PR DESCRIPTION
**Pull request by tapios**
## Purpose 

Tries to make ClimaAtmos unit tests more meaningful and comprehensive:
* Removes tests of ClimaCore (they should live in ClimaCore)
* Adds test of key computations, e.g., kinetic energy and strain rate in the presence of spherical geometry and topography, to provide more comprehensive coverage
* Replaces `@assert's` by `@test`
* Removes redundant tests
* Parallelizes unit tests into test groups executed simultaneously in CI

Also fixes two unrelated issues that caused errors/problems:
* Fixed `vertical_transport` → `vertical_advection` bug in advection.jl
* Made `thermo_vars` GPU-compatible via multiple dispatch in precomputed_quantities.jl
* Downgraded Thermodynamics compat entry

## To-do

Several other tests still need to be refactored/added. For example, we need conservation unit tests.

## Content

----
- [ ] I have read and checked the items on the review checklist.

### CODE REVIEW

File: test/parameterized_tendencies/microphysics/precipitation.jl [hunk 1]
```diff
@@ -148,156 +44,178 @@ end
...
+        ᶜYₜ = zero(Y)
+        (; turbconv_model, moisture_model, microphysics_model) = p.atmos
+
+        # Expected cache variables for 0-moment
+        expected_vars =
+            (:ᶜS_ρq_tot, :ᶜS_ρe_tot, :surface_rain_flux, :surface_snow_flux, :ᶜwₜqₜ, :ᶜwₕhₜ)
+        test_precipitation_setup!(Y, p, expected_vars)
+
+        # Source term should be near-zero for initial conditions
+        @test maximum(abs.(p.precomputed.ᶜS_ρq_tot)) <= sqrt(eps(FT))
```
Comment (szy21) [inline test/parameterized_tendencies/microphysics/precipitation.jl]: I don't think this test is necessary. It's ok to have a larger source term if e.g. we have a different initial condition.

File: test/parameterized_tendencies/microphysics/precipitation.jl [hunk 2]
```diff
@@ -148,156 +44,178 @@ end
...
+        CA.cloud_condensate_tendency!(
+            ᶜYₜ,
+            Y,
+            p,
+            moisture_model,
+            microphysics_model,
+            turbconv_model,
+        )
+        @test !any(isnan, ᶜYₜ.c.ρq_liq)
+        @test !any(isnan, ᶜYₜ.c.ρq_ice)
```
Comment (szy21) [inline test/parameterized_tendencies/microphysics/precipitation.jl:157]:
```suggestion
        @test !any(isnan, ᶜYₜ.c.ρq_ice)
        
        @test ᶜYₜ.c.ρq_tot == FT(0)
```

File: test/parameterized_tendencies/microphysics/precipitation.jl [hunk 3]
```diff
@@ -148,156 +44,178 @@ end
...
+        @test !any(isnan, ᶜYₜ.c.ρn_liq)
```
Comment (szy21) [inline test/parameterized_tendencies/microphysics/precipitation.jl:215]:
```suggestion
        @test !any(isnan, ᶜYₜ.c.ρn_liq)
        
        @test ᶜYₜ.c.ρq_tot == FT(0)
```

File: test/parameterized_tendencies/sponge/rayleigh_sponge.jl [hunk 4]
```diff
@@ -1,53 +1,120 @@
...
-
-    # Test when not using a Rayleigh sponge.
-    computed = CA.rayleigh_sponge_tendency_uₕ(ᶜuₕ, nothing)
-    @test computed isa NullBroadcasted
-    @. ᶜuₕ += computed # test that it can broadcast
+    ᶠz = Fields.coordinate_field(ᶠspace).z
+
+    # Create test velocity fields
+    ᶜuₕ = Fields.Field(Geometry.Covariant12Vector{FT}, ᶜspace)
+    fill!(parent(ᶜuₕ), FT(1))  # Both components = 1
```
Comment (szy21) [inline test/parameterized_tendencies/sponge/rayleigh_sponge.jl]: I was always told to avoid parent if I can, although I don't know exactly why.

File: test/parameterized_tendencies/sponge/rayleigh_sponge.jl [hunk 5]
```diff
@@ -1,53 +1,120 @@
...
+            FT(2) * sin(FT(π) / 2 * (ᶠz - z_damping) / (z_max - z_damping))^2,
+            FT(0),
+        )
+
+        @test CA.β_rayleigh_uₕ.(rs, ᶜz, z_max) ≈ expected_uₕ
+        @test CA.β_rayleigh_w.(rs, ᶠz, z_max) ≈ expected_w
+
+        # Test that damping is zero at z=0 (below z_damping)
+        @test CA.β_rayleigh_uₕ(rs, FT(0), z_max) == FT(0)
+        @test CA.β_rayleigh_w(rs, FT(0), z_max) == FT(0)
```
Comment (szy21) [inline test/parameterized_tendencies/sponge/rayleigh_sponge.jl]:
```suggestion

```
This is covered by previous lines?

File: test/parameterized_tendencies/sponge/rayleigh_sponge.jl [hunk 6]
```diff
@@ -1,53 +1,120 @@
...
+            α_w = FT(1),
+            α_sgs_tracer = FT(2),
+        )
+
+        # Tendency = -β * (χʲ - χ)
+        tendency = CA.rayleigh_sponge_tendency_sgs_tracer(ᶜχʲ, ᶜχ, rs)
+        β = CA.β_rayleigh_sgs_tracer.(rs, ᶜz, z_max)
+        expected = @. -β * (ᶜχʲ - ᶜχ)
+
+        # Since ᶜχʲ = 2 and ᶜχ = 1, the difference is 1
```
Comment (szy21) [inline test/parameterized_tendencies/sponge/rayleigh_sponge.jl]:
```suggestion
```

File: test/parameterized_tendencies/sponge/viscous_sponge.jl [hunk 7]
```diff
@@ -1,37 +1,86 @@
...
+        )
+
+        @test CA.β_viscous.(s, ᶜz, z_max) ≈ expected
+
+        # Test that damping is zero below z_damping
+        @test CA.β_viscous(s, FT(0), z_max) == FT(0)
+        @test CA.β_viscous(s, z_damping - FT(0.1), z_max) == FT(0)
+
+        # Test that damping is maximum at z_max (= κ₂)
+        @test CA.β_viscous(s, z_max, z_max) ≈ κ₂
```
Comment (szy21) [inline test/parameterized_tendencies/sponge/viscous_sponge.jl]:
```suggestion
```
I think it's a bit redundant, although I'm fine with it if you want to keep them.

File: test/coupler_compatibility.jl [hunk 8]
```diff
@@ -10,222 +10,18 @@ import ClimaAtmos.SurfaceConditions: projected_vector_data, CT1, CT2
...
-    sfc_T =
-        @. TD.air_temperature(thermo_params, p.precomputed.sfc_conditions.ts)
-    @test all(isequal(T1), parent(sfc_T))
-    update_surface_fields!(sfc_ts, sfc_conditions, FT(T2), p)
-    CA.SurfaceConditions.set_surface_conditions!(p, sfc_conditions, sfc_ts)
-    sfc_T =
-        @. TD.air_temperature(thermo_params, p.precomputed.sfc_conditions.ts)
-    @test all(isequal(T2), parent(sfc_T))
-end
+# TODO: Add surface flux exchange tests after new SurfaceFluxes.jl is integrated
```
Comment (szy21) [inline test/coupler_compatibility.jl:20]: The hacky version doesn't depend on SurfaceFluxes and is still working, although I'm not sure how useful it is.

File: test/parameter_tests.jl [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        @test params.turbconv_params isa CAP.TurbulenceConvectionParameters
+        @test params.microphysics_cloud_params isa NamedTuple
+        @test params.microphysics_cloud_params.liquid isa CAP.CM.Parameters.CloudLiquid
+
+        # Verify physical constants (sanity check for Earth, would need to be changed for other planets)
+        # R_d for dry air ~ 287 J/kg/K
+        @test CAP.R_d(params) ≈ FT(287.0) rtol = 0.01
+        # Planet radius ~ 6371 km
+        @test CAP.planet_radius(params) ≈ FT(6.371e6) rtol = 0.01
+        # Gravity ~ 9.8 m/s^2
```
Comment (szy21) [inline test/parameter_tests.jl]:
```suggestion
        # Gravity ~ 9.81 m/s^2
```

File: test/surface_albedo.jl [hunk 10]
```diff
@@ -1,193 +1,104 @@
...
+            [0.042499, 0.037779, 0.035150, 0.033099, 0.031358, 0.029819],
+            [0.064824, 0.059713, 0.056866, 0.054646, 0.052761, 0.051095],
+            [0.082365, 0.076947, 0.073930, 0.071576, 0.069578, 0.067811],
+        ]
+
+        cos_30 = FT(cosd(30))
+        for (i, n) in enumerate(FT.([1.2, 1.34, 1.45]))
+            albedo_model = ClimaAtmos.RegressionFunctionAlbedo{FT}(; n)
+            α_diffuse = ClimaAtmos.surface_albedo_diffuse(albedo_model)
+            computed = [α_diffuse(FT(0), cos_30, wind) for wind in FT.(0:5:25)]
```
Comment (szy21) [inline test/surface_albedo.jl]:
```suggestion
            computed = [α_diffuse(FT(0), FT(cosd(30)), wind) for wind in FT.(0:5:25)]
```

File: test/surface_albedo.jl [hunk 11]
```diff
@@ -1,193 +1,104 @@
...
+    @testset "Diffuse albedo (Figure 4)" begin
+        # Reference values from Jin et al 2011 Figure 4
+        # Rows: n = 1.2, 1.34, 1.45; Columns: wind = 0:5:25
+        reference = [
+            [0.042499, 0.037779, 0.035150, 0.033099, 0.031358, 0.029819],
+            [0.064824, 0.059713, 0.056866, 0.054646, 0.052761, 0.051095],
+            [0.082365, 0.076947, 0.073930, 0.071576, 0.069578, 0.067811],
+        ]
+
+        cos_30 = FT(cosd(30))
```
Comment (szy21) [inline test/surface_albedo.jl]:
```suggestion
```

File: test/test_helpers.jl [hunk 12]
```diff
@@ -158,6 +158,111 @@ function get_spherical_spaces(; FT = Float32)
...
+    device = ClimaComms.CPUSingleThreaded()
+    context = ClimaComms.SingletonCommsContext(device)
+
+    # Horizontal: cubed sphere
+    radius = FT(6.371e6)  # Earth radius
+    ne = 4  # elements per cube face edge
+    Nq = 4  # polynomial degree + 1
+    domain = Domains.SphereDomain(radius)
+    mesh = Meshes.EquiangularCubedSphere(domain, ne)
+    topology = Topologies.Topology2D(context, mesh)
```
Comment (szy21) [inline test/test_helpers.jl]:
```suggestion
    topology = Topologies.Topology2D(context, mesh, Topologies.spacefillingcurve(mesh))
```
(There is a bug that we are not using spacefillingcurve now, and we should fix it.)

File: test/test_helpers.jl [hunk 13]
```diff
@@ -158,6 +158,111 @@ function get_spherical_spaces(; FT = Float32)
...
+function get_spherical_extruded_spaces_with_topography(; FT = Float64)
+    context = ClimaComms.SingletonCommsContext()
+
+    # Horizontal: coarse cubed sphere
+    radius = FT(6.371e6)
+    ne = 4
+    Nq = 4
+    domain = Domains.SphereDomain(radius)
+    mesh = Meshes.EquiangularCubedSphere(domain, ne)
+    topology = Topologies.Topology2D(context, mesh)
```
Comment (szy21) [inline test/test_helpers.jl]:
```suggestion
    topology = Topologies.Topology2D(context, mesh, Topologies.spacefillingcurve(mesh))
```

File: test/utilities.jl [hunk 14]
```diff
@@ -65,158 +46,324 @@ end
...
+    _, fcoords = get_coords(cent_space, face_space)
+    ᶠu, ᶠv, ᶠw = taylor_green_ic(fcoords)
+    FT = eltype(ᶠu)
+    halflevel = ClimaCore.Utilities.half
+
+    # Method 1: horizontal_integral_at_boundary(field, level)
+    # Extracts the level internally and integrates
+    @test CA.horizontal_integral_at_boundary(ᶠu, halflevel) <= sqrt(eps(FT))
+    @test CA.horizontal_integral_at_boundary(ᶠv, halflevel) <= sqrt(eps(FT))
+    @test isfinite(CA.horizontal_integral_at_boundary(ᶠw, halflevel))  # w=0, just check it runs
```
Comment (szy21) [inline test/utilities.jl]:
```suggestion
    @test CA.horizontal_integral_at_boundary(ᶠw, halflevel))  == 0
```
or ≈0 if it doesn't work

File: test/utilities.jl [hunk 15]
```diff
@@ -65,158 +46,324 @@ end
...
+
+    # Method 2: horizontal_integral_at_boundary(level_field)
+    # Takes a pre-extracted 2D horizontal slice
+    ᶠuₛ = Fields.level(ᶠu, halflevel)
+    ᶠvₛ = Fields.level(ᶠv, halflevel)
+    ᶠwₛ = Fields.level(ᶠw, halflevel)
+    @test CA.horizontal_integral_at_boundary(ᶠuₛ) <= sqrt(eps(FT))
+    @test CA.horizontal_integral_at_boundary(ᶠvₛ) <= sqrt(eps(FT))
+    @test isfinite(CA.horizontal_integral_at_boundary(ᶠwₛ))
```
Comment (szy21) [inline test/utilities.jl]:
```suggestion
    @test CA.horizontal_integral_at_boundary(ᶠw, halflevel) == 0
```

File: test/variable_manipulations_tests.jl [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+    tc_model = MockTurbconvModel(0.5)
+
+    # Case 1: Large area fraction a = ρa/ρ = 0.8/1.0 = 0.8 > a_half (0.5)
+    # Weight should be > 0.5, result dominated by SGS value (8/0.8 = 10)
+    ρaχ = 8.0
+    ρa = 0.8
+    ρχ_fallback = 5.0
+    ρ_fallback = 1.0
+    val = CA.specific(ρaχ, ρa, ρχ_fallback, ρ_fallback, tc_model)
+    @test val ≈ 10.0 rtol = 0.1 # Should be close to SGS value
```
Comment (szy21) [inline test/variable_manipulations_tests.jl]:
```suggestion
    @test val ≈ ρaχ / ρa rtol = 0.1 # Should be close to SGS value
```

File: test/variable_manipulations_tests.jl [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+
+    # Case 2: Small area fraction a = 0.1 < a_half (0.5)
+    # Weight should be < 0.5, result blended with fallback (5/1 = 5)
+    ρaχ = 1.0
+    ρa = 0.1
+    val_small = CA.specific(ρaχ, ρa, ρχ_fallback, ρ_fallback, tc_model)
+    # Should be blended between 10 (SGS) and 5 (GridMean)
+    # 5.0 comes from specific(ρχ_fallback, ρ_fallback)
+    @test CA.specific(ρχ_fallback, ρ_fallback) < val_small < 10.0
```
Comment (szy21) [inline test/variable_manipulations_tests.jl]:
```suggestion
    @test CA.specific(ρχ_fallback, ρ_fallback) < val_small < ρaχ / ρa
```

File: test/parameterized_tendencies/sponge/rayleigh_sponge.jl (hunk 4 above)
Comment (tapios) [reply test/parameterized_tendencies/sponge/rayleigh_sponge.jl]: Ok. Replaced it with ClimaCore field broadcasting (but I'm not sure whether/why it's better).

File: test/coupler_compatibility.jl (hunk 8 above)
Comment (tapios) [reply test/coupler_compatibility.jl:20]: I rather have clear and meaningful tests once we are done refactoring the way boundary conditions are handled.
